### ISSUE TITLE
NullPointerException

### ISSUE THREAD

**Issue by sunnykinger** [Baseflow/flutter-permission-handler#609]
## 🐛 Bug Report

I am using crashlytics, and I have been receiving NullPointerException reports,

### Stacktrace
`Caused by java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.Map.containsKey(java.lang.Object)' on a null object reference
       at com.baseflow.permissionhandler.PermissionManager.onRequestPermissionsResult(PermissionManager.java:119)
       at io.flutter.embedding.engine.FlutterEngineConnectionRegistry$FlutterEngineActivityPluginBinding.onRequestPermissionsResult(FlutterEngineConnectionRegistry.java:712)
       at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.onRequestPermissionsResult(FlutterEngineConnectionRegistry.java:412)
       at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onRequestPermissionsResult(FlutterActivityAndFragmentDelegate.java:627)
       at io.flutter.embedding.android.FlutterActivity.onRequestPermissionsResult(FlutterActivity.java:648)
       at android.app.Activity.dispatchRequestPermissionsResult(Activity.java:8724)
       at android.app.Activity.dispatchActivityResult(Activity.java:8545)
       at android.app.ActivityThread.deliverResults(ActivityThread.java:5499)
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4851)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4901)
       at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
       at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2307)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8512)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1139)`

**Version:** 8.1.1 (library)

**Platform:** 
- [x ] :robot: Android

**Comment by JDDV** [comment]
Hi @sunnykinger, I do not know much about crashlytics, do you perhaps know or can you find out at what moment this might happen? And is it also possible to see if it only happens at certain brand/types of devices? I looked through the code but I don't see where a NullPointerException  might happen, and I never experienced that myself. I would love to hear some more information so we can figure this problem out!

**Comment by sunnykinger** [comment]
@JDDV Yes, I looked through the code as well, didn't find how NullPointerException came. The reports in Crashlytics came from Samsung, Vivo and Xiaomi.

**Comment by JDDV** [comment]
Hi @sunnykinger, Is it possible for you to share a code snippet where you ask for the permissions?
And is your application only for Android, or also for iOS?

I would like to see which permissions you request and how you request them to determine if this issue only occurs for certain permissions, or if it happens by the way how the permission is requested.

This might help us narrowing down the problem, thanks advance!

**Comment by windows7lake** [comment]
Hi, @JDDV , crashlytics report the same issue to me.
This bug was triggered after our users click the notification to open app. After checking my code, I have never invoked "requestPermissions" in code when click action was triggered. I assume that the plugin "permission_handler" was initing by flutter, and after that system send a permission request which will be recivied by `onRequestPermissionsResult` in `PermissionManager`("permission_handler").

**Comment by windows7lake** [comment]
``` dart
final class PermissionManager implements PluginRegistry.ActivityResultListener, PluginRegistry.RequestPermissionsResultListener {

    private Map<Integer, Integer> requestResults;


    @Override
    public boolean onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
         ........
            } else if (!requestResults.containsKey(permission)) {
                requestResults.put(
                        permission,
                        PermissionUtils.toPermissionStatus(this.activity, permissionName, result));
            }
         ........
        return true;
    }


    void requestPermissions(
            List<Integer> permissions,
            Activity activity,
            RequestPermissionsSuccessCallback successCallback,
            ErrorCallback errorCallback) {
         ........
        this.requestResults = new HashMap<>();
         ........
    }
```

**Comment by JDDV** [comment]
Hi @windows7lake, the crashlytics report, can you send me the stack trace maybe?
I need to know which permission is causing it, cause I've seen multiple issues now which requested different permission which gave a `NullPointerException`. In the stacktrace might be valuable information for that. I'd love to see your stack trace so I can help you out better. Thanks in advance!

**Comment by windows7lake** [comment]
```
Fatal Exception: java.lang.RuntimeException: Unable to resume activity {com.addcn.android.design591/com.addcn.android.design591.MainActivity}: java.lang.RuntimeException: Failure delivering result ResultInfo{who=@android:requestPermissions:, request=24, result=0, data=null} to activity {com.addcn.android.design591/com.addcn.android.design591.MainActivity}: java.lang.NullPointerException: Attempt to invoke interface method 'void f.d.a.n$b.a(java.util.Map)' on a null object reference
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3853)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3893)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3078)
       at android.app.ActivityThread.-wrap11(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1766)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:214)
       at android.app.ActivityThread.main(ActivityThread.java:6977)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:528)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:910)
Caused by java.lang.RuntimeException: Failure delivering result ResultInfo{who=@android:requestPermissions:, request=24, result=0, data=null} to activity {com.addcn.android.design591/com.addcn.android.design591.MainActivity}: java.lang.NullPointerException: Attempt to invoke interface method 'void f.d.a.n$b.a(java.util.Map)' on a null object reference
       at android.app.ActivityThread.deliverResults(ActivityThread.java:4621)
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3825)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3893)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3078)
       at android.app.ActivityThread.-wrap11(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1766)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:214)
       at android.app.ActivityThread.main(ActivityThread.java:6977)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:528)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:910)
Caused by java.lang.NullPointerException: Attempt to invoke interface method 'void f.d.a.n$b.a(java.util.Map)' on a null object reference
       at com.baseflow.permissionhandler.PermissionManager.onRequestPermissionsResult(PermissionManager.java:164)
       at io.flutter.embedding.engine.FlutterEngineConnectionRegistry$FlutterEngineActivityPluginBinding.onRequestPermissionsResult(FlutterEngineConnectionRegistry.java:712)
       at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.onRequestPermissionsResult(FlutterEngineConnectionRegistry.java:412)
       at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onRequestPermissionsResult(FlutterActivityAndFragmentDelegate.java:627)
       at io.flutter.embedding.android.FlutterActivity.onRequestPermissionsResult(FlutterActivity.java:648)
       at android.app.Activity.dispatchRequestPermissionsResult(Activity.java:7669)
       at android.app.Activity.dispatchActivityResult(Activity.java:7520)
       at android.app.ActivityThread.deliverResults(ActivityThread.java:4617)
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3825)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3893)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3078)
       at android.app.ActivityThread.-wrap11(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1766)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:214)
       at android.app.ActivityThread.main(ActivityThread.java:6977)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:528)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:910)

```

**Comment by windows7lake** [comment]
`ResultInfo{who=@android:requestPermissions:, request=24, result=0, data=null}`
As you can see, the request permission is empty, so I also difficult to know which permission was requested. It may be a bug from system, and null-safety may be helpful.

**Comment by skripsimu** [comment]
any update for this issues? i got the same error report from crashlytics

**Comment by PlugFox** [comment]
Same problem

```
Fatal Exception: java.lang.RuntimeException: Unable to resume activity {ru.ozon.magistral/dev.plugfox.magistral.MainActivity}: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=212, result=-1, data=null} to activity {ru.ozon.magistral/dev.plugfox.magistral.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.pm.PackageManager android.app.Activity.getPackageManager()' on a null object reference
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4907)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4944)
       at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
       at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2325)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8633)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
 ```

```
Caused by java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=212, result=-1, data=null} to activity {ru.ozon.magistral/dev.plugfox.magistral.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.pm.PackageManager android.app.Activity.getPackageManager()' on a null object reference
       at android.app.ActivityThread.deliverResults(ActivityThread.java:5579)
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4894)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4944)
       at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
       at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2325)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8633)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

```
Caused by java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.pm.PackageManager android.app.Activity.getPackageManager()' on a null object reference
       at com.baseflow.permissionhandler.PermissionManager.onActivityResult(PermissionManager.java:76)
       at io.flutter.embedding.engine.FlutterEngineConnectionRegistry$FlutterEngineActivityPluginBinding.onActivityResult(FlutterEngineConnectionRegistry.java:25)
       at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.onActivityResult(FlutterEngineConnectionRegistry.java:15)
       at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onActivityResult(FlutterActivityAndFragmentDelegate.java:51)
       at io.flutter.embedding.android.FlutterActivity.onActivityResult(FlutterActivity.java:10)
       at android.app.Activity.dispatchActivityResult(Activity.java:8550)
       at android.app.ActivityThread.deliverResults(ActivityThread.java:5572)
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4894)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4944)
       at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
       at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2325)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8633)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

**Comment by JDDV** [comment]
@windows7lake and @sunnykinger  do you request the `requestInstallPackages` permission? I tried the example app from the plugin and request the `requestInstallPackages permission. It turns out that Android (running on SDK 30. Android 11) closes the app when the intent comes up where you switch between approving or denying the permission (see image below here to see which screen I talk about).

![Screenshot_1639653923](https://user-images.githubusercontent.com/47666699/146363561-dc30df42-586f-4caf-9b8d-28a02da07a67.png)

This behavior of closing the app does not occur in Android 12 and Android 10 or lower, this only happens in Android 11 and as far I can see is an Android bug, not a permission_handler bug.
Can you, @windows7lake, @sunnykinger, @PlugFox tell me or test if this is the  case for your app too? and can you check in the crashlytics if this is just like with @PlugFox chrashlytics mostly an Android 11 problem? I would like to gather more information on this problem so we can solve it. 

So far I think either the `requestInstallPackages` permission is causing it, or the `activity` variable is not yet set somehow, and when doing something like calling `activity.getPackageManager` will result in a NullPointerException.

I woud love to see some code snippets about how you ask the permissions so I have more information to work with.

**Comment by windows7lake** [comment]
![image](https://user-images.githubusercontent.com/10578898/146702361-47b66e6f-b1c5-4254-aadf-b0fdc6a4cf27.png)
@JDDV  as you can see, firebase report that it's not only crash on Android 11. And in my project, I have never used `requestInstallPackages` permission.
I think null check with `map` may help to aviod this issue.

**Comment by JDDV** [comment]
Hi @sunnykinger and @windows7lake. I looked into this issue a bit further and it looks like the `NullPointerException` that got thrown should be solved since version `^8.1.6`. The issue was that in the `PermissionManager` class the variable `requestResults` could be null and when the `containsKey` was called on the `requestResults` variable it would throw a `NullPointerException`. Since version `8.1.6` this issue is solved by adding a null check on `requestResults`.

@PlugFox who also mentioned a NullPointerException had a different problem. You can read more in issue #758.

**Comment by ShishirXT** [comment]
Hi @JDDV 
We're encountering a similar issue. In the release version of the application, we're receiving the following exception.

**Package Version: permission_handler: ^11.0.0
Flutter Version: 3.13.6**


```
Exception java.lang.RuntimeException:
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:4975)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:5018)
  at android.app.servertransaction.ResumeActivityItem.execute (ResumeActivityItem.java:60)
  at android.app.servertransaction.ActivityTransactionItem.execute (ActivityTransactionItem.java:45)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:176)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:97)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2374)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loopOnce (Looper.java:233)
  at android.os.Looper.loop (Looper.java:344)
  at android.app.ActivityThread.main (ActivityThread.java:8249)
  at java.lang.reflect.Method.invoke
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:589)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1071)
Caused by java.lang.RuntimeException:
  at android.app.ActivityThread.deliverResults (ActivityThread.java:5568)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:4953)
Caused by java.lang.NullPointerException: Attempt to invoke interface method 'java.lang.Object java.util.Map.put(java.lang.Object, java.lang.Object)' on a null object reference
  at com.baseflow.permissionhandler.PermissionManager.onActivityResult (PermissionManager.java)
  at io.flutter.embedding.engine.FlutterEngineConnectionRegistry$FlutterEngineActivityPluginBinding.onActivityResult
  at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.onActivityResult
  at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onActivityResult
  at io.flutter.embedding.android.FlutterActivity.onActivityResult
  at android.app.Activity.dispatchActivityResult (Activity.java:8472)
  at android.app.ActivityThread.deliverResults (ActivityThread.java:5561)

Exception : java.lang.NullPointerException	Attempt to invoke interface method 'java.lang.Object java.util.Map.put(java.lang.Object, java.lang.Object)' on a null object reference
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:5094)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:5127)
  at android.app.servertransaction.ResumeActivityItem.execute (ResumeActivityItem.java:54)
  at android.app.servertransaction.ActivityTransactionItem.execute (ActivityTransactionItem.java:45)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:176)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:97)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2441)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loopOnce (Looper.java:233)
  at android.os.Looper.loop (Looper.java:334)
  at android.app.ActivityThread.main (ActivityThread.java:8396)
  at java.lang.reflect.Method.invoke
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:582)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1068)
Caused by java.lang.RuntimeException:
  at android.app.ActivityThread.deliverResults (ActivityThread.java:5651)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:5081)
Caused by java.lang.NullPointerException: Attempt to invoke interface method 'java.lang.Object java.util.Map.put(java.lang.Object, java.lang.Object)' on a null object reference
  at com.baseflow.permissionhandler.PermissionManager.onActivityResult (PermissionManager.java)
  at io.flutter.embedding.engine.FlutterEngineConnectionRegistry$FlutterEngineActivityPluginBinding.onActivityResult
  at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.onActivityResult
  at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onActivityResult
  at io.flutter.embedding.android.FlutterActivity.onActivityResult
  at android.app.Activity.dispatchActivityResult (Activity.java:8718)
  at android.app.ActivityThread.deliverResults (ActivityThread.java:5644)


```

Impacted users are using the following Android Version

<img width="597" alt="Screenshot 2024-04-22 at 1 10 22 PM (1)" src="https://github.com/Baseflow/flutter-permission-handler/assets/159430338/dac6d84a-2d05-43ef-96a1-924896b75bf8">

**Since we're unable to replicate it, we need your assistance.**
